jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs

```diff
@@ -0,0 +1,137 @@
+use std::sync::Arc;
+
+pub struct KeyService {
+    idp_configs: HashMap<Box<str>, OpenIdConfig>, // <issuer (`iss`), OpenIdConfig>
+}
+
+#[allow(unused)]
+impl KeyService {
+    /// initializes a new `KeyService` with the provided OpenID configuration endpoints.
+    ///
+    /// this method fetches the OpenID configuration and the associated keys (JWKS) for each
+    /// endpoint, populating the internal `idp_configs` map. any HTTP errors or parsing
+    /// failures will return a corresponding `Error`.
+    pub fn new(openid_conf_endpoints: Vec<&str>) -> Result<Self, Error> {
+        let mut idp_configs = HashMap::new();
+
+        // fetch IDP configs
+        for endpoint in openid_conf_endpoints {
+            // TODO: use a reusable client when making requests
+            let conf_src: OpenIdConfigSource = get(endpoint)
+                .map_err(Error::HttpError)?
+                .error_for_status()
+                .map_err(Error::HttpError)?
+                .json()
+                .map_err(Error::RequestDeserializationError)?;
+            let (issuer, conf) = OpenIdConfig::from_source(conf_src);
+            idp_configs.insert(issuer, conf);
+        }
+
+        /// retrieves a decoding key based on the provided key ID (`kid`).
+        ///
+        /// this method first attempts to retrieve the key from the local key store. if the key
+        /// is not found, it will refresh the JWKS and try again. if the key is still not found,
+        /// an error of type `KeyNotFound` is returned.
+        for (iss, conf) in &mut idp_configs {
+            let jwks: JwkSet = get(&*conf.jwks_uri)
+                .map_err(Error::HttpError)?
+                .error_for_status()
+                .map_err(Error::HttpError)?
+                .json()
+                .map_err(Error::RequestDeserializationError)?;
+            let mut decoding_keys = conf.decoding_keys.write().unwrap();
+            for jwk in jwks.keys {
+                let decoding_key = DecodingKey::from_jwk(&jwk).map_err(Error::KeyParsingError)?;
+                let key_id = jwk.common.key_id.ok_or(Error::MissingKeyId)?;
+                decoding_keys.insert(key_id.into(), Arc::new(decoding_key));
+            }
+        }
+
+        Ok(Self { idp_configs })
+    }
+
+    /// retrieves a decoding key based on the provided key ID (`kid`).
+    ///
+    /// this method first attempts to retrieve the key from the local key store. if the key
+    /// is not found, it will refresh the JWKS and try again. if the key is still not found,
+    /// an error of type `KeyNotFound` is returned.
+    pub fn get_key(&self, kid: &str) -> Result<Arc<DecodingKey>, Error> {
+        for (iss, config) in &self.idp_configs {
+            // first try to get the key from the local keystore
+            if let Some(key) = self.get_key_from_iss(&iss, &kid) {
+                return Ok(key.clone());
+            } else {
+                println!("could not find {}, updating jwks", kid);
```

Comment:
It's better to use `eprintln!`

Reply:
resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)